Abins 0.1

dymkowsk opened the pull request:
2nd pull request for ABINS project.

This pull request is a continuation of the pull request #17529 , which ran into some merge issues with master. HENCE TO REVIEW THIS PULL REQUEST ALSO SEE COMMENTS ON #17529.

**To test:**

<!-- Instructions for testing. -->

<!-- RELEASE NOTES
Either edit the file in docs/source/release/... and it will be in your pull request or state
*Does not need to be in the release notes.*
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

Anders-Markvardsen commented:
Please specify files you don't want to be picked up by https://github.com/mantidproject/mantid/blob/master/Testing/SystemTests/tests/analysis/LoadLotsOfFiles.py

Anders-Markvardsen commented:
System tests, unit tests, usage example and docs added. Code reviewed and interface tested. A good new addition to mantid. :shipit:

quantumsteve commented:
@mantidproject/gatekeepers Please hold off on merging this until we discuss packaging 3rd party projects at tomorrow's TSC meeting.

Anders-Markvardsen commented:
Note this work has been validated by several scientists included isis inelastic group members Sanghamitra and Steward. Further, it has been benchmarked

[benchmark.pdf](https://github.com/mantidproject/mantid/files/713415/benchmark.pdf)

and this work has been presented SCD/ISIS on several occasions with various input.

In my testing I have focussed on ensuring the right level of testing etc is in place.

quantumsteve commented:
@Anders-Markvardsen Could Abins be packaged as a separate python package that depends on Mantid? For example check out [PyVDrive](https://github.com/neutrons/PyVDrive), [FastGR](https://github.com/neutrons/FastGR), and [mslice](https://github.com/mantidproject/mslice).

Anders-Markvardsen commented:
@quantumsteve the request was to have abins fully tested and integrated with mantid. Speak later, I will come along for the first 1/2 hour of the tsc meeting today

martyngigg commented:
@Anders-Markvardsen @quantumsteve I think I may have misunderstood. I can't see anything GUI-related here so this wouldn't fit in that style of approach.

* 62 new files are test data references
* 18 new file are tests themselves
* and then it's basically a set of python modules and a Workflow algorithm.

My only one comment here really would be that the algorithm names don't match our standards with all caps for `ABINS`.

peterfpeterson commented:
Even a "calculation" only addition can be done as an external package. My reluctance to the addition  is partially maintenance and partially that it is not what [mantid is about](https://www.mantidproject.org/Main_Page).

martyngigg commented:
@peterfpeterson Indeed it can but this doesn't give users very easy access to it. How would you suggest they would go about using it? If this is going to be part of the workflow for indirect how is it different to any of the other scripts etc?

rosswhitfield commented:
Since this is a lot of new python code can we make sure it's compatible with python 3 before it's merged. At the very least add the `__future__` imports to all the python files as shown on https://www.mantidproject.org/Python_3

martyngigg commented:
@dymkowsk I believe you are in the process of adding the `__future__` imports, is that right?

dymkowsk commented:
yep I will add it asap.

rosswhitfield commented:
Why didn't you add `absolute_import` with the `__future__` imports?

dymkowsk commented:
I skipped absolut_import because if I add it modules from AbinsModules are invisible for Mantid. Martyn said that it is OK for now to skip it (but I should fix it for the next release).

rosswhitfield commented:
But that would imply they also don't work with python 3! I think it should be added and the problem fixed.

dymkowsk commented:
Ok, 
As a fast solution I was thinking about adding  absolut_import to __future__ and the following lines (or equivalent) to each file

import mantid
one_path = mantid.config["pythonscripts.directories"].split(";")[0]
abins_path = os.path.join(one_path[:one_path.index("scripts")], "scripts", "AbinsModules") mantid.config["pythonscripts.directories"] = mantid.config["pythonscripts.directories"] + ";" + abins_path

Is it Ok, If I add something like that to each file?

rosswhitfield commented:
You can fix the imports by for example
```
diff --git a/scripts/AbinsModules/__init__.py b/scripts/AbinsModules/__init__.py
index effff4dda0..0cab899206 100644
--- a/scripts/AbinsModules/__init__.py
+++ b/scripts/AbinsModules/__init__.py
@@ -17,31 +17,32 @@ algorithms and data objects that are:
 
 """
 # flake8: noqa
-from IOmodule import IOmodule
+from __future__ import (absolute_import, division, print_function)
+from .IOmodule import IOmodule
 
 # Frequency generator
-from FrequencyPowderGenerator import FrequencyPowderGenerator
+from .FrequencyPowderGenerator import FrequencyPowderGenerator
 
 # Loading modules
-from GeneralDFTProgram import GeneralDFTProgram
-from LoadCASTEP import LoadCASTEP
-from LoadCRYSTAL import LoadCRYSTAL
+from .GeneralDFTProgram import GeneralDFTProgram
+from .LoadCASTEP import LoadCASTEP
+from .LoadCRYSTAL import LoadCRYSTAL
 
 # Calculating modules
-from CalculatePowder import CalculatePowder
-from CalculateSingleCrystal import CalculateSingleCrystal
-from CalculateDWSingleCrystal import CalculateDWSingleCrystal
-from CalculateS import CalculateS
+from .CalculatePowder import CalculatePowder
+from .CalculateSingleCrystal import CalculateSingleCrystal
+from .CalculateDWSingleCrystal import CalculateDWSingleCrystal
+from .CalculateS import CalculateS
 
 # Data
-from GeneralData import GeneralData
-from KpointsData import KpointsData
-from AtomsData import AtomsDaTa
-from AbinsData import AbinsData
-from DWSingleCrystalData import DWSingleCrystalData
-from SingleCrystalData import SingleCrystalData
-from PowderData import PowderData
-from SData import SData
+from .GeneralData import GeneralData
+from .KpointsData import KpointsData
+from .AtomsData import AtomsDaTa
+from .AbinsData import AbinsData
+from .DWSingleCrystalData import DWSingleCrystalData
+from .SingleCrystalData import SingleCrystalData
+from .PowderData import PowderData
+from .SData import SData
 
 # Instruments
-from InstrumentProducer import InstrumentProducer
+from .InstrumentProducer import InstrumentProducer
```

dymkowsk commented:
Ok, will try it. Thanks!

martyngigg commented on Framework/PythonInterface/test/python/plugins/algorithms/AbinsAdvancedParametersTest.py:
```diff
@@ -0,0 +1,301 @@
+    def setUp(self):
+
+        # set up input for Abins
+        self._Si2 = "Si2-sc_AbinsAdvancedParameters"
```
I think this needs to be updated to `Si2-sc_ABINSAdvancedParameters` as the filename has not changed

martyngigg commented on Framework/PythonInterface/test/python/plugins/algorithms/AbinsBasicTest.py:
```diff
@@ -0,0 +1,209 @@
+class AbinsBasicTest(unittest.TestCase):
+
+    _si2 = "Si2-sc_Abins"
+    _squaricn = "squaricn_sum_Abins"
```
Same as above `Abins` to `ABINS`

martyngigg commented on Framework/PythonInterface/test/python/plugins/algorithms/AbinsBasicTest.py:
```diff
@@ -0,0 +1,209 @@
+        @return:
+        """
+        Abins(DFTprogram=self._dft_program,
+              PhononFile="benzene_Abins.phonon",
```
Again, `Abins`-> `ABINS`

dymkowsk commented:
Yes I have already done this:

       new file:   Testing/Data/UnitTest/Si2-sc_Abins.phonon.md5
	new file:   Testing/Data/UnitTest/Si2-sc_AbinsAdvancedParameters.phonon.md5
	new file:   Testing/Data/UnitTest/benzene_Abins.dat.md5
	new file:   Testing/Data/UnitTest/benzene_Abins.phonon.md5
	new file:   Testing/Data/UnitTest/squaricn_sum_Abins.phonon.md5

There are also missing two blank lines in GeneralData.py which I have also added.

 now I am trying to absolut_import. I have also  I modified my __init_ files and it worked. Now I am modifying my other files. I will push all modyfications together.

Anders-Markvardsen commented on scripts/AbinsModules/Instruments/__init__.py:
```diff
@@ -1,5 +1,13 @@
-from TwoDMap import TwoDMap
-from Instrument import Instrument
+from __future__ import (absolute_import, division, print_function)
+# Make sure we can find Instruments...
```
I must admit I am new to absolute_import, but @dymkowsk just checking if you don't include this work around will absolute_import not work?

martyngigg replied:
@Anders-Markvardsen You're correct, they should not be required. https://docs.python.org/2.5/whatsnew/pep-328.html gives a good overview of this.

martyngigg commented on scripts/AbinsModules/FrequencyPowderGenerator.py:
```diff
@@ -0,0 +1,107 @@
+
+import numpy as np
+
+import AbinsConstants
```
Lines such as this will fail with `absolute_import`. They can be fixed by slightly rewriting the line to say:
```
from . import AbinsConstants
```

This will need to be done in the remaining places after the removal of the `sys.path` updates that @Anders-Markvardsen has already mentioned.

martyngigg commented on scripts/AbinsModules/__init__.py:
```diff
@@ -0,0 +1,60 @@
+"""
```
This looks like it has been copied from the main `mantid` file?

It should instead be a docstring about the `AbinsModules`. Can we get this updated?

dymkowsk replied:
Sure, no problem. I will apply changes to imports and documentation asap.

martyngigg replied:
Thank you.

martyngigg commented:
@dymkowsk Thanks for those latest updates. We're nearly there. I ran the tests under Python 3 and a couple more issues came up that wouldn't be covered by this. We're going to be adding Python 3 checks soon fixing them would be helpful.

I have created a `.patch` file at https://gist.github.com/martyngigg/74089aff29369e0ab300c1a6c1731a9a that fixes all of the remaining test failures under Python 3. You can apply it using the following command, which must be run from the root of the repository:

```
curl -sSL https://gist.githubusercontent.com/martyngigg/74089aff29369e0ab300c1a6c1731a9a/raw/278db0f5742f496d6cebee3f00a3c0f4688a7892/abins-py3.patch | git apply
```

Can you apply this and push to your fork's branch?

martyngigg commented:
@dymkowsk I would suggest applying the above command **before** merging master to fix the conflicts.

dymkowsk commented:
Thank you for patch. I will apply it later today after lunch.

rosswhitfield commented:
@martyngigg did you mean to disable with `xtest` some of the tests in your patch?

martyngigg commented:
@rosswhitfield Thank you for pointing out my carelessness. I didn't mean to do that 😔

martyngigg commented:
@dymkowsk could you fix the test methods that are named `xtest` and change them back to `test`

martyngigg commented:
Thanks
